File: wallet_core/platform_support/ios/PlatformSupport/AttestedKey/Bridge/AttestedKey.swift
//
//  AttestedKey.swift
//  PlatformSupport
//
//  Created by The Wallet Developers on 24/10/2024.
//

import CryptoKit
import DeviceCheck
import Foundation

final class AttestedKey {
    private static let appAttest = DCAppAttestService.shared
}

extension AttestedKey: AttestedKeyBridge {
    func keyType() -> AttestedKeyType {
        .apple
    }

    func generate() async throws(AttestedKeyError) -> String {
        do {
            return try await Self.appAttest.generateKey()
        } catch let error as DCError {
            throw AttestedKeyError.from(error)
        } catch {
            fatalError(error.localizedDescription)
        }
    }

    func attest(identifier: String, challenge: [UInt8], googleCloudProjectId: UInt64) async throws(AttestedKeyError) -> AttestationData {
        do {
            let attestation = try await Self.appAttest.attestKey(identifier, clientDataHash: Data(challenge))

            return .apple(attestationData: Array(attestation))
        } catch let error as DCError {
            throw AttestedKeyError.from(error)
        } catch {
            fatalError(error.localizedDescription)
        }
    }

    func sign(identifier: String, payload: [UInt8]) async throws(AttestedKeyError) -> [UInt8] {
        let clientDataHash = Data(SHA256.hash(data: Data(payload)))

        do {
            let assertion = try await Self.appAttest.generateAssertion(identifier, clientDataHash: clientDataHash)
            return Array(assertion)
        } catch let error as DCError {
            throw AttestedKeyError.from(error)
        } catch {
            fatalError(error.localizedDescription)
        }
    }

    func publicKey(identifier _: String) throws(AttestedKeyError) -> [UInt8] {
        // Retrieving the public key is only supported as part of key attestation for iOS.
        // This method is only implemented for Android and should not be called on this platform.
        throw .MethodUnimplemented
    }

    func delete(identifier _: String) throws(AttestedKeyError) {
        // Deleting an attested key is not supported by iOS.
        // This method is only implemented for Android and should not be called on this platform.
        throw .MethodUnimplemented
    }
}

File: wallet_core/platform_support/src/bridge/attested_key.rs
use std::fmt::Debug;

use async_trait::async_trait;

use super::get_platform_support;

/// Implementation of `AttestedKeyError` from the UDL file.
#[derive(Debug, thiserror::Error)]
pub enum AttestedKeyError {
    #[error("key/app attestation is not supported on this device")]
    AttestationNotSupported,
    #[error("the called method is not impemented for this platform")]
    MethodUnimplemented,
    #[error("vendor server is unreachable: {details}")]
    ServerUnreachable { details: String },
    #[error("{reason}")]
    Other { reason: String },
}

/// Implementation of `AttestedKeyType` from the UDL file.
pub enum AttestedKeyType {
    Apple,
    Google,
}

/// Implementation of `AttestationData` from the UDL file.
pub enum AttestationData {
    Apple {
        attestation_data: Vec<u8>,
    },
    Google {
        certificate_chain: Vec<Vec<u8>>,
        app_attestation_token: String,
    },
}

/// Implementation of `AttestedKeyBridge` from the UDL file.
// Unfortunately we cannot use the built-in support for traits with async methods,
// as those are not object safe, so we have to rely on the async_trait crate instead.
#[async_trait]
pub trait AttestedKeyBridge: Send + Sync + Debug {
    fn key_type(&self) -> AttestedKeyType;
    async fn generate(&self) -> Result<String, AttestedKeyError>;
    async fn attest(
        &self,
        identifier: String,
        challenge: Vec<u8>,
        google_cloud_project_number: u64,
    ) -> Result<AttestationData, AttestedKeyError>;
    async fn sign(&self, identifier: String, payload: Vec<u8>) -> Result<Vec<u8>, AttestedKeyError>;

    // Only supported on Android
    async fn public_key(&self, identifier: String) -> Result<Vec<u8>, AttestedKeyError>;
    async fn delete(&self, identifier: String) -> Result<(), AttestedKeyError>;
}

/// Convenience function to access the a reference to `AttestedKeyBridge`,
/// as set by by the native implementation.
pub fn get_attested_key_bridge() -> &'static dyn AttestedKeyBridge {
    get_platform_support().attested_key.as_ref()
}
